test/solver_tests/snj_solver_test.py

```diff
@@ -0,0 +1,642 @@
+        cm_lineage = pd.DataFrame.from_dict(
+            {
+                "a": [1, 1, 0],
+                "b": [1, 2, 0],
+                "c": [1, 2, 1],
+                "d": [2, 0, 0],
+                "e": [2, 0, 2],
+            },
+            orient="index",
+            columns=["x1", "x2", "x3"],
+        )
+
+        self.tree_lineage = cas.data.CassiopeiaTree(character_matrix=cm_lineage)
+
+        # ------------- CM with Duplicates -----------------------
+        cm_dupe = pd.DataFrame.from_dict(
+            {
+                "a": [1, 1, 0],
+                "b": [1, 2, 0],
+                "c": [1, 2, 1],
+                "d": [2, 0, 0],
+                "e": [2, 0, 2],
+                "f": [2, 0, 2],
+            },
+            orient="index",
+            columns=["x1", "x2", "x3"],
+        )
+
+        self.tree_dupe = cas.data.CassiopeiaTree(
+            character_matrix=cm_dupe
+        )
+
+        # ------------- SNJ with modified hamming dissimilarity ------------
+        priors = {0: {1: 0.5, 2: 0.5}, 1: {1: 0.2, 2: 0.8}, 2: {1: 0.3, 2: 0.7}}
+        self.tree_lineage_priors = cas.data.CassiopeiaTree(
+            character_matrix=cm_lineage, priors=priors
+        )
+
+    def test_constructor(self):
+        self.assertIsNotNone(self.snj_solver.dissimilarity_function)
+        self.assertIsNotNone(self.tree_general.get_dissimilarity_map())
+
+        nothing_solver = cas.solver.SpectralNeighborJoiningSolver(
+            dissimilarity_function=None, add_root=False
+        )
+
+        no_root_tree = cas.data.CassiopeiaTree(
+            character_matrix=self.cm_general,
+            dissimilarity_map=self.sim_general,
+        )
+
+        with self.assertRaises(cas.solver.DistanceSolver.DistanceSolverError):
+            nothing_solver.solve(no_root_tree)
+
+        no_root_solver = cas.solver.SpectralNeighborJoiningSolver(
+            dissimilarity_function=None, add_root=True
+        )
+
+        with self.assertRaises(cas.solver.DistanceSolver.DistanceSolverError):
+            no_root_solver.solve(no_root_tree)
+
+        root_only_tree = cas.data.CassiopeiaTree(
+            character_matrix=self.cm_general, root_sample_name="b"
+        )
+
+        with self.assertRaises(cas.solver.DistanceSolver.DistanceSolverError):
+            nothing_solver.solve(root_only_tree)
+
+        snj_solver_fn = cas.solver.SpectralNeighborJoiningSolver(
+            add_root=True 
+        ) 
+        snj_solver_fn.solve(self.tree_general)
+
+        self.assertEqual(
+            self.tree_general.get_dissimilarity_map().loc["a", "b"], 1/3
+        )
+
+    def test_compute_lambda_pairwise(self):
+        self.lambda_pairwise = [[i] for i in range(self.sim_general.values.shape[0])]
+
+        # run to generate similarity map
+        self.snj_solver.get_dissimilarity_map(self.tree_lineage)
+
+        svd2_vals = self.snj_solver.compute_lambda(self.lambda_pairwise)
+
+        expected_q = pd.DataFrame.from_dict(
```

Comment:
Please standardize these variable names for similar outputs throughout. For example, here it's called `expected_q`, but later is called `expected_lambda` and `expected_lambda_matrix`.